### PR TITLE
Feature/applications

### PR DESCRIPTION
**Pull request by govTechSatish**
1. Get Application
2. Search Applications
3. View Application

### CODE REVIEW

File: app/views/applications.njk [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          }
+        }
+      ],
+      rows: applications
+    }) }}
+    {{ pagination.pager(page, totalPages, applications.length, total, url) }}
+  </div>
+  {% endif %}
+  
+{% endblock %}
```
Comment (st3v3nhunt) [inline app/views/applications.njk:103]: Newline please.

File: docker-compose.override.yaml [hunk 2]
```diff
@@ -1,17 +1,27 @@
...
+
+  redis:
+    image: redis:4.0.14
+    volumes:
+      - redis_data:/data
+    ports:
+      - "6380:6379"
+
+volumes:
+  redis_data:
```
Comment (st3v3nhunt) [inline docker-compose.override.yaml:27]: new line please.

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
Looking good, a few questions and suggestions added inline and some more general ones here:

- Please update the `README`, it is just the default atm
- The paging control isn't working. `Next` is going to page 4 rather than then next page and `Previous` is going to page 1 rather than the previous page. Might be best to remove it from this initial PR and it as a feature on a subsequent PR
- It might be worth moving `pagination` into a `lib` directory rather than it sitting in the root of `app`, it'll help with organisation
- Have a link to `applications` on the home page would be useful

File: app/routes/applications.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+const currentPath = `/${viewTemplate}`
+const { getApplications } = require('../messaging/applications')
+const { getPagination, getPagingData } = require('../pagination')
+const Joi = require('joi')
+const { setAppSearch, getAppSearch } = require('../session')
+const keys = require('../session/keys')
+
+async function createModel (request, errorMessage) {
+  const { limit, offset } = getPagination(request.query.page, request.query.limit)
+  const page = request.query.page ?? 1
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/routes/applications.js:11]: Good Catch.

File: docker-compose.override.yaml (hunk 2 above)
Comment (govTechSatish) [reply docker-compose.override.yaml:27]: Done

File: app/views/applications.njk (hunk 1 above)
Comment (govTechSatish) [reply app/views/applications.njk:103]: Done

File: app/routes/applications.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  const path = request.headers.path ?? ''
+  const searchText = getAppSearch(request, keys.appSearch.searchText)
+  const searchType = getAppSearch(request, keys.appSearch.searchType)
+  const apps = await getApplications(searchType ?? '', searchText ?? '', limit, offset, request.yar.id)
+  if (apps.applicationState !== 'not_found') {
+    const pagingData = getPagingData(apps.total ?? 0, limit, page, path)
+
+    let statusClass = 'govuk-tag--grey'; let status = 'Pending'
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/routes/applications.js:19]: Done

File: app/routes/applications.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/routes/applications.js:11]: Done.

File: app/routes/applications.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
+const viewTemplate = 'applications'
+const currentPath = `/${viewTemplate}`
+const { getApplications } = require('../messaging/applications')
+const { getPagination, getPagingData } = require('../pagination')
+const Joi = require('joi')
+const { setAppSearch, getAppSearch } = require('../session')
+const keys = require('../session/keys')
+
+async function createModel (request, errorMessage) {
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/routes/applications.js:9]: Done

File: app/plugins/router.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const routes = [].concat(
+  require('../routes/assets'),
+  require('../routes/accessibility'),
+  require('../routes/cookies'),
+  require('../routes/healthy'),
+  require('../routes/healthz'),
+  require('../routes/home'),
+  require('../routes/applications'),
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/plugins/router.js]: Done

File: app/pagination.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+const getPagination = (page = 1, limit = 10) => {
+  const offset = page === 1 ? 0 : (page - 1) * limit
+  return { limit, offset }
+}
+
+const getPagingData = (total, limit, page, url) => {
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/pagination.js]: done

File: app/pagination.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const getPagination = (page = 1, limit = 10) => {
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/pagination.js]: done.

File: app/messaging/send-message.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+const { MessageSender } = require('ffc-messaging')
+const createMessage = require('./create-message')
+
+const sendMessage = async (body, type, config, options) => {
+  const message = createMessage(body, type, options)
+  console.log('sending message', message)
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/messaging/send-message.js]: Removed log

File: app/frontend/src/css/index.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import { initAll } from 'govuk-frontend'
```
[the remark this answers is not in the document]
Comment (govTechSatish) [reply app/frontend/src/css/index.js]: removed.

**Review by st3v3nhunt — APPROVED** [review]

**Pull request closed without merging**